### PR TITLE
test: update integration test

### LINKED ISSUE #67: Add complete integration test for model-backend
Test highlights (including but not limited to):
- [ ] CV task parser
- [ ] use a tiny dummy model for testing, no need to download remote model

### LINKED ISSUE #69: `make test` relies on a giant online yolov4 model
When we do 
```bash
make all
make test
```
The tests won't pass because the tests rely on an existing online yolov4 model.

We should use a tiny dummy model for testing and fix this issue to make the tests self-contained.

### PR DESCRIPTION
**Pull request by Phelan164**
Because

- need integration test to make sure feature work correctly

This commit

- add integration test for model-backend and update for pipeline-backend
- closes #67 
- closes #69

Limitation: error code is not covered in this PR. It will be added during working [this ticket](https://github.com/instill-ai/model-backend/issues/33)

Note: docker image for model-backend is dev which should update when have a new release

### CODE REVIEW

File: tests/integration-tests/pipeline-backend-rest.js [hunk 1]
```diff
@@ -20,6 +23,52 @@ export let options = {
...
+          [`PATCH ${apiHost}/models/${model_name}/versions/1 det response Versions`]: (r) =>
+          r.json().versions.length > 0,       
+          [`PATCH ${apiHost}/models/${model_name}/versions/1 det response Version 1 Status`]: (r) =>
+          r.json().versions[0].status === "ONLINE",  
+      });  
+    });
+  } 
 }
 
 const dogImg = open(`${__ENV.TEST_FOLDER_ABS_PATH}/dog.jpg`, "b");
```
Comment (xiaofei-du) [inline tests/integration-tests/pipeline-backend-rest.js:74]:
```suggestion
const dogImg = open(`${__ENV.TEST_FOLDER_ABS_PATH}/tests/integration-tests/data/dog.jpg`, "b");
```

File: tests/integration-tests/pipeline-backend-rest.js [hunk 2]
```diff
@@ -12,6 +12,9 @@ import * as pipelineConstants from "./pipeline-backend-constants.js";
 
 const apiHost = "http://127.0.0.1:8446";
 
+const model_name = "yolov4"
```
Comment (xiaofei-du) [inline tests/integration-tests/pipeline-backend-rest.js:15]:
```suggestion
const model_name = pipelineConstants.detectionModel.name;
```

File: tests/integration-tests/pipeline-backend-rest.js (hunk 1 above)
Comment (xiaofei-du) [reply tests/integration-tests/pipeline-backend-rest.js:74]: In `pipeline-backend-rest.js`, change `group("Pipelines API: Trigger a pipeline with classification model", () => {` to `    group("Pipelines API: Trigger a pipeline", () => {`

File: tests/integration-tests/model-backend-rest.js [hunk 3]
```diff
@@ -31,11 +37,649 @@ export default function (data) {
...
+        headers: genHeader(`application/json`),
+      }), {
+        "DELETE clean up response Status": (r) =>
+          r.status === 200 // TODO: update status to 201
+      }); 
+      check(http.request("DELETE", `${apiHost}/models/${model_name2}`, null, {
+        headers: genHeader(`application/json`),
+      }), {
+        "DELETE clean up response Status": (r) =>
+          r.status === 200 // TODO: update status to 201
```
Comment (xiaofei-du) [inline tests/integration-tests/model-backend-rest.js:120]:
```suggestion
          r.status === 200 // TODO: update status to 204
```

File: tests/integration-tests/model-backend-rest.js [hunk 4]
```diff
@@ -31,11 +37,649 @@ export default function (data) {
...
+          r.json().CVTask === "UNDEFINED",   
+          "POST /models (multipart) cvtask undefined response Versions": (r) =>
+          r.json().Versions.length === 1,                                        
+      });        
+      // clean up
+      check(http.request("DELETE", `${apiHost}/models/${model_name1}`, null, {
+        headers: genHeader(`application/json`),
+      }), {
+        "DELETE clean up response Status": (r) =>
+          r.status === 200 // TODO: update status to 201
```
Comment (xiaofei-du) [inline tests/integration-tests/model-backend-rest.js:114]:
```suggestion
          r.status === 200 // TODO: update status to 204
```

File: tests/integration-tests/model-backend-rest.js [hunk 5]
```diff
@@ -31,11 +37,649 @@ export default function (data) {
...
+      }); 
+      check(http.request("DELETE", `${apiHost}/models/${model_name3}`, null, {
+        headers: genHeader(`application/json`),
+      }), {
+        "DELETE clean up response Status": (r) =>
+          r.status === 200 // TODO: update status to 201
```
Comment (xiaofei-du) [inline tests/integration-tests/model-backend-rest.js:126]:
```suggestion
          r.status === 200 // TODO: update status to 204
```

File: tests/integration-tests/model-backend-rest.js [hunk 6]
```diff
@@ -31,11 +37,649 @@ export default function (data) {
...
+          "PATCH /models (multipart) cvtask cls response Version 2 Status": (r) =>
+          r.json().versions[1].status === "OFFLINE",                                                            
+      });       
+      
+      // clean up
+      check(http.request("DELETE", `${apiHost}/models/${model_name}`, null, {
+        headers: genHeader(`application/json`),
+      }), {
+        "DELETE clean up response Status": (r) =>
+          r.status === 200 // TODO: update status to 201
```
Comment (xiaofei-du) [inline tests/integration-tests/model-backend-rest.js:340]:
```suggestion
          r.status === 200 // TODO: update status to 204
```

File: tests/integration-tests/model-backend-rest.js [hunk 7]
```diff
@@ -31,11 +37,649 @@ export default function (data) {
...
+          [`GET ${apiHost}/models/${model_name} full_Name`]: (r) =>
+          r.json().full_Name === `local-user/${model_name}`,                                
+      });      
+      
+      // clean up
+      check(http.request("DELETE", `${apiHost}/models/${model_name}`, null, {
+        headers: genHeader(`application/json`),
+      }), {
+        "DELETE clean up response Status": (r) =>
+          r.status === 200 // TODO: update status to 201
```
Comment (xiaofei-du) [inline tests/integration-tests/model-backend-rest.js:620]:
```suggestion
          r.status === 200 // TODO: update status to 204
```

**Review by xiaofei-du — CHANGES_REQUESTED** [review]
I have left some comments. Two more questions:

1. Can we delete all the models in the `teardown` function in `model-backend-rest.js` like deleting all pipelines in `pipeline-backend-rest.js`. This can make sure the model-backend tests won't affect other tests.

2. In `model-backend-rest.js`, I see that we have covered tests for `CLASSIFICATION` and `DETECTION` 👍 . Can we also add similar tests to cover creating a model without specifying any CV task and making inferences with this model? I think it will make the basic tests more complete.

**Comment by Phelan164** [comment]
@xiaofei-du 
>Can we delete all the models in the teardown function in model-backend-rest.js like deleting all pipelines in pipeline-backend-rest.js. This can make sure the model-backend tests won't affect other tests.

In each test case, I always delete the created model. Why could not delete in teardown? Because many models created during testing (model name is random string) then need delete in each test case to prevent impact to other test cases (example list model will be return much more, etc)

**Comment by Phelan164** [comment]
>In model-backend-rest.js, I see that we have covered tests for CLASSIFICATION and DETECTION 👍 . Can we also add similar tests to cover creating a model without specifying any CV task and making inferences with this model? I think it will make the basic tests more complete.

yes, agree. It missing. I will add that

File: docker-compose.yml [hunk 8]
```diff
@@ -26,7 +26,7 @@ services:
 
   model_backend_migrate:
     container_name: model-backend-migrate
-    image: instill/model-backend:v0.2.0-alpha
+    image: instill/model-backend:dev
```
Comment (pinglin) [inline docker-compose.yml:29]: @Phelan164 We use the release version in vdp docker-compose.

File: docker-compose.yml [hunk 9]
```diff
@@ -40,7 +40,7 @@ services:
 
   model_backend:
     container_name: model-backend
-    image: instill/model-backend:v0.2.0-alpha
+    image: instill/model-backend:dev
```
Comment (pinglin) [inline docker-compose.yml:43]: @Phelan164 For the same reason as the above comment, please drop this change.

**Review by pinglin — CHANGES_REQUESTED** [review]

**Comment by xiaofei-du** [comment]
> 



> @xiaofei-du
> 
> > Can we delete all the models in the teardown function in model-backend-rest.js like deleting all pipelines in pipeline-backend-rest.js. This can make sure the model-backend tests won't affect other tests.
> 
> In each test case, I always delete the created model. Why could not delete in teardown? Because many models created during testing (model name is random string) then need delete in each test case to prevent impact to other test cases (example list model will be return much more, etc)

Yeah, we should delete the created resource in each test case. But when other people contribute tests, they may forget it. So filling the `teardown` process is just to add the extra security.

**Review by xiaofei-du — APPROVED** [review]
LGTM 👍

**Review by pinglin — APPROVED** [review]

**Review by bochengyang — APPROVED** [review]
LGTM 🚀

**Comment by bochengyang** [comment]
close #50

**Pull request merged**